dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/protocol/AbstractInvoker.java

```diff
@@ -151,6 +149,14 @@ public Result invoke(Invocation inv) throws RpcException {
         }
         RpcUtils.attachInvocationIdIfAsync(getUrl(), invocation);
 
+        // If we have attachment, refresh the attachment to ensure that the last invoke's attach does not affect this call.
+        // In most cases, attachments contains interface, timeout, group, and token.
+        // These info should be refreshed in each invoke.
+        //
+        // See https://github.com/apache/incubator-dubbo/issues/2981
+        if (attachment != null && attachment.size() > 0) {
+            invocation.addAttachments(attachment);
```

Comment:
Just curious to know, as comment indicates refreshing timeout, group, and token. Do we need any place to be also refresh other than invocation?

Reply:
Hey, maybe the description here is a bit problematic. We are not refreshing but covering. I will modify it later.

The global attachment here is information such as interface and timeout obtained from the URL. We want to make sure that the interface and timeout used for each call are in the url of the current invoker, not the other way. Other ways such as: RpcContext.getContext.setAttachment
So, before invoke, we should override these info.
:)

Reply:
What if user has overridden the `timeout` or `token` value in RpcContext/Invocation, if you put replace `addAttachmentsIfAbsent ` with `addAttachments` and put it at the last place,  the user will never has a chance to override the keys listed here https://github.com/apache/incubator-dubbo/blob/e07038b00e057bb3cd6989252534b581039916f9/dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/protocol/AbstractInvoker.java#L78

Reply:
Cannot we just remove  or replace the INTERFACE_KEY with the right one?

Reply:
I even find that the INTERFACE_KEY useless on the wire.  Instead `path` was used instead:
https://github.com/apache/incubator-dubbo/blob/e07038b00e057bb3cd6989252534b581039916f9/dubbo-rpc/dubbo-rpc-dubbo/src/main/java/org/apache/dubbo/rpc/protocol/dubbo/DubboProtocol.java#L215

Reply:
https://github.com/apache/incubator-dubbo/blob/33f1726713a4c5f2ce5880613acf2320ae2a4502/dubbo-rpc/dubbo-rpc-api/src/main/java/org/apache/dubbo/rpc/filter/ContextFilter.java#L46

I am simply thinking of remove INTERFACE_KEY in here. GROUP_KEY, TOKEN_KEY and TIMEOUT_KEY would not have this problem because they've already been removed.

Reply:
Ok!
i haven't notice it yet...

Reply:
yes.
This problem does happen in the process of using Thrift.